R/explanation.R

```diff
@@ -0,0 +1,330 @@
+#' of test observations, and \code{p} equals the total number of features.
+#'
+#' @export
+#'
+#' @author Camilla Lingjaerde
+explain <- function(x, explainer, approach, prediction_zero, ...) {
+
+  # Check input for x
+  if (!is.matrix(x) & !is.data.frame(x)) {
+    stop("x should be a matrix or a dataframe.")
+  }
+
+  # Check input for approach
+  str_error <- paste(
+    "It seems that you passed a non-valid value for approach.",
+    "It should be either 'empirical', 'gaussian', 'copula' or",
+    "a list."
+  )
+
+  nms_valid <- c("empirical", "gaussian", "copula", "combined")
+
+  if (is.list(approach)) {
+    if (!all(colnames(approach) %in% nms_valid)) {
+      stop(str_error)
+    }
+    class(x) <- "combined"
+  } else {
+    if (!is.element(approach, nms_valid)) {
+      stop(str_error)
+    }
+    class(x) <- approach
+  }
+
+  UseMethod("explain", x)
+}
+
+#' @param type String or list. Only applicable when \code{approach='empirical'}. If a string, the
+#' type of empirical approach to use,  equal to 'independence, 'gaussian' or 'fixed_sigma'. If a
+#' list, the elements in the list refers to the rows in \code{x} that ought to be included in
+#' each of the empirical approaches.
+#'
+#' @param fixed_sigma_vec Vector or numeric. Only applicable when \code{approach='empirical'} and
+#' \code{type='fixed_sigma'}. The bandwidth to use. Default value \code{0.1}
+#'
+#' @param AICc_no_samp_per_optim Positive integer. Only applicable when
+#' \code{approach='empirical'} and \code{type='AICc_each_k'} or
+#' \code{type='AICc_full'}. Number of samples to consider in AICc optimization.
+#'
+#' @param AIC_optim_max_eval Positive integer. Only applicable when \code{approach='empirical'}
+#' and \code{type='AICc_each_k'} or \code{type='AICc_full'}. Numeric. Maximum value when
+#' optimizing the AICc.
+#'
+#' @param AIC_optim_startval Numeric. Only applicable when \code{approach='empirical'} and
+#' \code{type='AICc_each_k'} or \code{type='AICc_full'}. Starting value when optimizing the AICc.
+#'
+#' @param w_threshold Postive integer between 0 and 1.
+#'
+#' @rdname explain
+#' @name explain
+#'
+#' @export
+explain.empirical <- function(x, explainer, approach, prediction_zero,
+                              type = "fixed_sigma", fixed_sigma_vec = 0.1,
+                              AICc_no_samp_per_optim = 1000, AIC_optim_max_eval = 20,
+                              AIC_optim_startval = 0.1, w_threshold = 0.95, ...) {
+
+  # Add arguments to explainer object
+  explainer$x_test <- as.matrix(x)
+  explainer$approach <- approach
+  explainer$type <- type
+  explainer$fixed_sigma_vec <- fixed_sigma_vec
+  explainer$AICc_no_samp_per_optim <- AICc_no_samp_per_optim
+  explainer$AIC_optim_max_eval <- AIC_optim_max_eval
+  explainer$AIC_optim_startval <- AIC_optim_startval
+  explainer$w_threshold <- w_threshold
+  explainer$n_samples <- AICc_no_samp_per_optim
```

Comment:
I think this might be wrong. The number of samples used in the AICc optimization need not be the same as the number of samples that we use for the empirical method in the end... They may very well be two differnet numbers. For now I don't manage to see how explainer$n_samples gets passed for the empirical method though. Please take look at this. Otherwise it looks good to me for now.

Reply:
You're right. I'll fix it.

Reply:
Fixed with 11773f4.

Reply:
Perfect!